### PR TITLE
fix MPP-2305: make inbound_call respond 200 with <Say>

### PR DESCRIPTION
**Pull request by groovecoder**
This PR fixes MPP-2305.

How to test:
1. Set your relay number to "blocking"
2. Call your relay number
   * [ ] You should hear "Sorry, that number is not available."

- [ ] I've added a unit test to test for potential regressions of this bug.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: api/views/phones.py [hunk 1]
```diff
@@ -477,7 +488,13 @@ def inbound_call(request):
 
     relay_number, real_phone = _get_phone_objects(inbound_to)
 
-    _check_disabled(relay_number, "calls")
+    number_disabled = _check_disabled(relay_number, "calls")
+    if number_disabled:
+        say = "Sorry, that number is not available."
+        return response.Response(
+            {"say": say}, status=200, template_name="twiml_blocked.xml"
```
Comment (jwhitlock) [inline api/views/phones.py:495]: This worked!

File: api/views/phones.py [hunk 2]
```diff
@@ -445,10 +444,18 @@ def inbound_sms(request):
...
         )
 
-    _check_disabled(relay_number, "texts")
+    number_disabled = _check_disabled(relay_number, "texts")
+    if number_disabled:
+        say = "Sorry, that number is not accepting texts."
+        return response.Response(
+            {"say": say}, status=200, template_name="twiml_blocked.xml"
+        )
+    _check_remaining(relay_number, "texts")
```
Comment (jwhitlock) [inline api/views/phones.py:458]: This `_check_remaining` call is a duplicate of line 442

File: api/views/phones.py [hunk 3]
```diff
@@ -445,10 +444,18 @@ def inbound_sms(request):
...
+            template_name="twiml_empty_response.xml",
         )
 
-    _check_disabled(relay_number, "texts")
+    number_disabled = _check_disabled(relay_number, "texts")
+    if number_disabled:
+        say = "Sorry, that number is not accepting texts."
+        return response.Response(
+            {"say": say}, status=200, template_name="twiml_blocked.xml"
+        )
```
Comment (jwhitlock) [inline api/views/phones.py:457]: This is OK but misleading. The `<Say>` command does nothing for incoming text messages, so there is no response. What about the `twiml_empty_response.xml` instead?

File: api/views/phones.py [hunk 4]
```diff
@@ -462,12 +469,16 @@ def inbound_sms(request):
     relay_number.remaining_texts -= 1
     relay_number.texts_forwarded += 1
     relay_number.save()
-    return response.Response(status=201, data={"message": "Relayed message to user."})
+    return response.Response(
+        {"message": "Relayed message to user."},
```
Comment (jwhitlock) [inline api/views/phones.py:473]: Is the `message` needed, so that _something_ is returned?

**Review by jwhitlock — CHANGES_REQUESTED** [review]
Thanks @groovecoder! The Voice blocked call works, but I think the SMS path could use some cleanup.

These response do _not_ change, but could in a future PR:

* The `RelayNumber` has hit monthly limits
* The `InboundContact` is blocked

File: api/views/phones.py (hunk 4 above)
Comment (groovecoder) [reply api/views/phones.py:473]: Not needed. Removed.

File: api/templates/twiml_error.xml [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (jwhitlock) [inline api/templates/twiml_error.xml:1]: Nit: this template is currently unused - response is constructed in code - so this file should be deleted

File: api/views/phones.py [hunk 6]
```diff
@@ -40,8 +40,7 @@
 from ..exceptions import ConflictError
 from ..permissions import HasPhoneService
 from ..renderers import (
-    TwilioInboundCallXMLRenderer,
-    TwilioInboundSMSXMLRenderer,
+    TemplateTwiMLRenderer,
```
Comment (jwhitlock) [inline api/views/phones.py:43]: 👍 good job unifying the renderers!

**Review by jwhitlock — APPROVED** [review]
Looks good, thanks @groovecoder! Feel free to change or ignore nits, I have a feeling we'll be touching this code again soon.

Here's the tested responses (*Updated for commit [c21796e](https://github.com/mozilla/fx-private-relay/pull/2626/commits/c21796e74b109bd7597762dca34431a9703d5071)*)

* With Google Voice `+12005551234`
* With Relay Number `+13025551234`

## Enabled
### Phone call
`201 Created`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Dial callerId="+12005551234">
        <Number>+13025551234</Number>
    </Dial>
</Response>
```

Voice call is connected. Voice status callback POST, `CallStatus=completed` at end of call.

### SMS

`201 Created`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response/>
```

SMS is relayed.

## Number `+12005551234` blocked
### Voice
`400 Bad Request`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <status_code>400</status_code>
    <code>invalid</code>
    <title>Number Is Not Accepting Calls.</title>
</Error>
```
Voice feedback (female) is "We're sorry, an application error has occurred. Good-bye.". Voice status callback gets POST, `CallStatus=completed`.

### SMS
`400 Bad Request`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <status_code>400</status_code>
    <code>invalid</code>
    <title>Number Is Not Accepting Texts.</title>
</Error>
```
No message is relayed.

## Relay number is disabled
### Voice

`200 OK`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Say>
        Sorry, that number is not available.
    </Say>
</Response>
```

Voice feedback (male) is "Sorry that number is not available.". Voice status callback POST with `CallStatus=completed`.

### SMS
`200 OK`
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Response/>
```
No SMS is relayed.

File: api/views/phones.py [hunk 7]
```diff
@@ -445,10 +444,18 @@ def inbound_sms(request):
...
+            {"message": "Sent reply to last sender."},
+            status=200,
+            template_name="twiml_empty_response.xml",
         )
 
-    _check_disabled(relay_number, "texts")
+    number_disabled = _check_disabled(relay_number, "texts")
+    if number_disabled:
+        return response.Response(
+            {"message": "Sorry, that number is not accepting texts."},
```
[the remark this answers is not in the document]
Comment (groovecoder) [reply api/views/phones.py:455]: And there was another unneeded `message` on L447!

File: api/templates/twiml_error.xml (hunk 5 above)
Comment (groovecoder) [reply api/templates/twiml_error.xml:1]: Hrm ... I think I really wanted to make `TemplateTwiMLRenderer.render` use this template, but IIRC I ran into a snag getting that done and probably forgot to finish it. I updated `TemplateTwiMLRenderer.render` to use it now.

**Review by jwhitlock — APPROVED** [review]
Thanks @groovecoder, the new code to use `twiml_error.xml` works! I updated my earlier comment with the new responses, which include a `<status_code>400</status_code>` element.

**Pull request merged**